### ISSUE TITLE
Soft-failures make federated bans racy and frustrate redaction

### ISSUE THREAD

**Issue by heftig** [matrix-org/synapse#9329]
I've come across this issue twice in two days:

- A user joins over federation and spams messages into the room
- Mjolnir bans the user and redacts their messages
- Users from other homeservers report that an unredacted message remains, but my homeserver doesn't have it

I think the issue is:

- Synapse soft-fails a message from the banned user that other homeservers have accepted
- ⇒ Redaction of this message is now impossible

A sample is below. Mjolnir's FirstMessageIsImageProtection banned and redacted the user after they sent three images, but their fourth image is soft-failed, making redaction impossible.

archlinux.org rejected the message:

```
[root@matrix ~]# journalctl --since='3 hours ago' | grep "\$WM5mW7VqMvW19ceHPt1r-jrLCPJE8pGCVXfVZKQ47io"
Feb 04 14:04:17 matrix.archlinux.org synapse[56716]: synapse.handlers.federation: [PUT-159337-$WM5mW7VqMvW19ceHPt1r-jrLCPJE8pGCVXfVZKQ47io] handling received PDU: <FrozenEventV3 event_id='$WM5mW7VqMvW19ceHPt1r-jrLCPJE8pGCVXfVZKQ47io', type='m.room.message', state_key=None>
Feb 04 14:04:17 matrix.archlinux.org synapse[56716]: synapse.state: [PUT-159337-$WM5mW7VqMvW19ceHPt1r-jrLCPJE8pGCVXfVZKQ47io] Resolving state for !SEgsRQLScqPxYtucHl:archlinux.org with groups [2601779, 2601782]
Feb 04 14:04:18 matrix.archlinux.org synapse[56716]: synapse.handlers.federation: [PUT-159337-$WM5mW7VqMvW19ceHPt1r-jrLCPJE8pGCVXfVZKQ47io] Soft-failing <FrozenEventV3 event_id='$WM5mW7VqMvW19ceHPt1r-jrLCPJE8pGCVXfVZKQ47io', type='m.room.message', state_key=None> because 403: User [bad-user] not in room !SEgsRQLScqPxYtucHl:archlinux.org (<FrozenEventV3 event_id='$SdIQbg5dST1Yly1f-_G8qERVvbOUPeO5P2lHt3Fe3vI', type='m.room.member', state_key='[bad-user]'>)
Feb 04 14:04:18 matrix.archlinux.org synapse[56716]: synapse.http.client: [PUT-159337-$WM5mW7VqMvW19ceHPt1r-jrLCPJE8pGCVXfVZKQ47io] Received response to POST http://127.0.0.1:9093/_synapse/replication/fed_send_events/cxkgfQOzda: 200
```

aria-net.org accepted the message:
```
:~$ grep "\$WM5mW7VqMvW19ceHPt1r-jrLCPJE8pGCVXfVZKQ47io" /var/log/syslog
Feb  4 14:04:17 akademeia matrix-synapse-worker_federation_reader_1[189517]: 2021-02-04 14:04:17,116 - synapse.handlers.federation - 189 - INFO - PUT-19046-$WM5mW7VqMvW19ceHPt1r-jrLCPJE8pGCVXfVZKQ47io - handling received PDU: <FrozenEventV3 event_id='$WM5mW7VqMvW19ceHPt1r-jrLCPJE8pGCVXfVZKQ47io', type='m.room.message', state_key=None>
Feb  4 14:04:17 akademeia matrix-synapse-worker_federation_reader_1[189517]: 2021-02-04 14:04:17,822 - synapse.http.client - 425 - INFO - PUT-19046-$WM5mW7VqMvW19ceHPt1r-jrLCPJE8pGCVXfVZKQ47io - Received response to POST http://127.0.0.1:8101/_synapse/replication/fed_send_events/cItXxahwLx: 200
```

### Version information

- **Homeserver**: matrix.archlinux.org
- **Version**: `{"server_version":"1.26.0","python_version":"3.9.1"}`
- **Install method**: pip
- **Platform**: Arch Linux

**Comment by erikjohnston** [comment]
This is a very interesting failure mode, and I'm not sure how best to deal with it.

To clarify: it's impossible to redact because the client never receives the event? (As opposed to the redact request failing?)

**Comment by heftig** [comment]
> To clarify: it's impossible to redact because the client never receives the event? (As opposed to the redact request failing?)

Correct. The clients on my homeserver don't see the event that needs to be redacted. Clearing caches doesn't help.

**Comment by richvdh** [comment]
related: #7744, #6539

**Comment by richvdh** [comment]
(though a different problem to those two)

**Comment by erikjohnston** [comment]
Some potential mitigations to this include:

1. Having a grace period of e.g. 30s (based on origin server ts) before we start soft failing events, to give time for the updated state to propagate.
2. "Backdating" soft fails, e.g. if we get a ban after receiving messages from the banned user Synapse could retroactively redact those events. This should be limited in some way, e.g. by timestamps? This is slightly complicated by the fact there's no easy way for the server to tell clients to redact local copies of events without sending an real redaction event.
3. Have a "moderation view", i.e. send down soft failed events to moderators so that they can see and react to those events. This might be confusing, and so messages that are soft failed should be marked as such. We could have a sync filter that allowed clients to opt into it?

We probably need to think about this at the spec level too.

**Comment by heftig** [comment]
> 1. Having a grace period of e.g. 30s (based on origin server ts) before we start soft failing events, to give time for the updated state to propagate.

I think this and 3 would also need improvements to Mjolnir's auto-redaction so it redacts post-ban events?

> 2. "Backdating" soft fails, e.g. if we get a ban after receiving messages from the banned user Synapse could retroactively redact those events. This should be limited in some way, e.g. by timestamps? This is slightly complicated by the fact there's no easy way for the server to tell clients to redact local copies of events without sending an real redaction event.

Would this use the event graph? I.e. if a server receives a ban event, events from this user that are not ancestors of this ban need to be "locally redacted"? And this redaction is limited by time?

**Comment by heftig** [comment]
> 1. Having a grace period of e.g. 30s (based on origin server ts) before we start soft failing events, to give time for the updated state to propagate.

We recently banned about a dozen bots and Mjolnir auto-redacted a couple hundred messages, and it took almost an hour to get all bans federated to matrix.org. I don't think the grace period will work when the room slows down due to a coordinated attack, when proper redaction is needed most.
